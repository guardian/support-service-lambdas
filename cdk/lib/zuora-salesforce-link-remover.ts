import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda';
import { aws_cloudwatch, Duration } from 'aws-cdk-lib';
import type { App } from 'aws-cdk-lib';
import {
	Alarm,
	Metric,
	Stats,
	TreatMissingData,
} from 'aws-cdk-lib/aws-cloudwatch';
import { SnsAction } from 'aws-cdk-lib/aws-cloudwatch-actions';
import { Rule, Schedule } from 'aws-cdk-lib/aws-events';
import { SfnStateMachine } from 'aws-cdk-lib/aws-events-targets';
import { Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Architecture, Runtime } from 'aws-cdk-lib/aws-lambda';
import { Topic } from 'aws-cdk-lib/aws-sns';
import {
	Choice,
	Condition,
	DefinitionBody,
	JsonPath,
	Map,
	Pass,
	StateMachine,
} from 'aws-cdk-lib/aws-stepfunctions';
import { LambdaInvoke } from 'aws-cdk-lib/aws-stepfunctions-tasks';

export class ZuoraSalesforceLinkRemover extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const appName = 'zuora-salesforce-link-remover';

		const allowPutMetric = new PolicyStatement({
			effect: Effect.ALLOW,
			actions: ['cloudwatch:PutMetricData'],
			resources: ['*'],
		});

		const getSalesforceBillingAccountsLambda = new GuLambdaFunction(
			this,
			'get-billing-accounts-lambda',
			{
				app: appName,
				functionName: `${appName}-get-billing-accounts-${this.stage}`,
				runtime: Runtime.NODEJS_20_X,
				environment: {
					Stage: this.stage,
				},
				handler: 'getBillingAccounts.handler',
				fileName: `${appName}.zip`,
				architecture: Architecture.ARM_64,
				initialPolicy: [
					new PolicyStatement({
						actions: ['secretsmanager:GetSecretValue'],
						resources: [
							`arn:aws:secretsmanager:${this.region}:${this.account}:secret:DEV/Salesforce/ConnectedApp/AwsConnectorSandbox-oO8Phf`,
							`arn:aws:secretsmanager:${this.region}:${this.account}:secret:DEV/Salesforce/User/integrationapiuser-rvxxrG`,
							`arn:aws:secretsmanager:${this.region}:${this.account}:secret:PROD/Salesforce/ConnectedApp/BillingAccountRemover-WUdrKa`,
							`arn:aws:secretsmanager:${this.region}:${this.account}:secret:PROD/Salesforce/User/BillingAccountRemoverAPIUser-UJ1SwZ`,
						],
					}),
					allowPutMetric,
				],
			},
		);

		const updateZuoraBillingAccountLambda = new GuLambdaFunction(
			this,
			'update-zuora-billing-account-lambda',
			{
				app: appName,
				functionName: `${appName}-update-zuora-billing-account-${this.stage}`,
				runtime: Runtime.NODEJS_20_X,
				environment: {
					Stage: this.stage,
				},
				handler: 'updateZuoraBillingAccount.handler',
				fileName: `${appName}.zip`,
				architecture: Architecture.ARM_64,
				initialPolicy: [
					new PolicyStatement({
						actions: ['secretsmanager:GetSecretValue'],
						resources: [
							`arn:aws:secretsmanager:${this.region}:${this.account}:secret:CODE/Zuora-OAuth/SupportServiceLambdas-S8QM4l`,
							`arn:aws:secretsmanager:${this.region}:${this.account}:secret:PROD/Zuora-OAuth/SupportServiceLambdas-Iu3KIT`,
						],
					}),
					allowPutMetric,
				],
			},
		);

		const updateSfBillingAccountsLambda = new GuLambdaFunction(
			this,
			'update-sf-billing-accounts-lambda',
			{
				app: appName,
				functionName: `${appName}-update-sf-billing-accounts-${this.stage}`,
				runtime: Runtime.NODEJS_20_X,
				environment: {
					Stage: this.stage,
				},
				handler: 'updateSfBillingAccounts.handler',
				fileName: `${appName}.zip`,
				architecture: Architecture.ARM_64,
				initialPolicy: [
					new PolicyStatement({
						actions: ['secretsmanager:GetSecretValue'],
						resources: [
							`arn:aws:secretsmanager:${this.region}:${this.account}:secret:DEV/Salesforce/ConnectedApp/AwsConnectorSandbox-oO8Phf`,
							`arn:aws:secretsmanager:${this.region}:${this.account}:secret:DEV/Salesforce/User/integrationapiuser-rvxxrG`,
							`arn:aws:secretsmanager:${this.region}:${this.account}:secret:PROD/Salesforce/ConnectedApp/BillingAccountRemover-WUdrKa`,
							`arn:aws:secretsmanager:${this.region}:${this.account}:secret:PROD/Salesforce/User/BillingAccountRemoverAPIUser-UJ1SwZ`,
						],
					}),
					allowPutMetric,
				],
			},
		);

		const getSalesforceBillingAccountsFromLambdaTask = new LambdaInvoke(
			this,
			'Get Salesforce Billing Accounts',
			{
				lambdaFunction: getSalesforceBillingAccountsLambda,
				outputPath: '$.Payload',
			},
		);

		const updateZuoraBillingAccountLambdaTask = new LambdaInvoke(
			this,
			'Update Zuora Billing Account',
			{
				lambdaFunction: updateZuoraBillingAccountLambda,
				outputPath: '$.Payload',
			},
		);

		const updateSfBillingAccountsLambdaTask = new LambdaInvoke(
			this,
			'Update Salesforce Billing Accounts',
			{
				lambdaFunction: updateSfBillingAccountsLambda,
				inputPath: '$.billingAccountProcessingAttempts',
				outputPath: '$.Payload',
			},
		);

		const billingAccountsProcessingMap = new Map(
			this,
			'Billing Accounts Processor Map',
			{
				maxConcurrency: 10,
				itemsPath: JsonPath.stringAt('$.billingAccountsToProcess'),
				itemSelector: {
					item: JsonPath.stringAt('$$.Map.Item.Value'),
				},
				resultPath: '$.billingAccountProcessingAttempts',
			},
		);

		const billingAccountsProcessingMapDefinition =
			updateZuoraBillingAccountLambdaTask;

		billingAccountsProcessingMap.itemProcessor(
			billingAccountsProcessingMapDefinition,
		);

		const billingAccountsExistChoice = new Choice(
			this,
			'Billing Accounts exist for processing?',
		)
			.when(
				Condition.isPresent('$.billingAccountsToProcess[0]'),
				billingAccountsProcessingMap.next(updateSfBillingAccountsLambdaTask),
			)
			.otherwise(new Pass(this, 'No Billing Accounts to process'));

		const definitionBody = DefinitionBody.fromChainable(
			getSalesforceBillingAccountsFromLambdaTask.next(
				billingAccountsExistChoice,
			),
		);

		const stateMachine = new StateMachine(
			this,
			`zuora-salesforce-link-remover-state-machine-${this.stage}`,
			{
				definitionBody: definitionBody,
			},
		);

		const cronEveryHour = { minute: '0', hour: '*' };
		const cronOncePerYear = { minute: '0', hour: '0', day: '1', month: '1' };
		const executionFrequency =
			this.stage === 'PROD' ? cronEveryHour : cronOncePerYear;

		new Rule(this, 'ScheduleStateMachineRule', {
			schedule: Schedule.cron(executionFrequency),
			targets: [new SfnStateMachine(stateMachine)],
			enabled: true,
		});

		const topic = Topic.fromTopicArn(
			this,
			'Topic',
			`arn:aws:sns:${this.region}:${this.account}:alarms-handler-topic-${this.stage}`,
		);

		const lambdaFunctions = [
			getSalesforceBillingAccountsLambda,
			updateZuoraBillingAccountLambda,
			updateSfBillingAccountsLambda,
		];

		lambdaFunctions.forEach((lambdaFunction, index) => {
			const alarm = new Alarm(this, `alarm-${index}`, {
				alarmName: `Zuora <-> Salesforce link remover - ${lambdaFunction.functionName} - something went wrong - ${this.stage}`,
				alarmDescription:
					'Something went wrong when executing the zuora <-> salesforce link remover. See Cloudwatch logs for more information on the error.',
				datapointsToAlarm: 1,
				evaluationPeriods: 1,
				actionsEnabled: true,
				comparisonOperator:
					aws_cloudwatch.ComparisonOperator.GREATER_THAN_THRESHOLD,
				metric: new Metric({
					metricName: 'Errors',
					namespace: 'AWS/Lambda',
					statistic: Stats.SUM,
					period: Duration.seconds(60),
					dimensionsMap: {
						FunctionName: lambdaFunction.functionName,
					},
				}),
				threshold: 0,
				treatMissingData: TreatMissingData.MISSING,
			});
			alarm.addAlarmAction(new SnsAction(topic));
		});
	}
}
