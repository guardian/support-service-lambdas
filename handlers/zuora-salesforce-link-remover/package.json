{
	"name": "zuora-salesforce-link-remover",
	"description": "A state machine that queries Salesforce for Zuora records that should be deleted for GDPR purposes, and updates the respective Billing Accounts in Zuora by removing the crmId.",
	"scripts": {
		"test": "jest --group=-integration",
		"it-test": "jest --group=integration",
		"build": "esbuild --bundle --platform=node --target=node20 --outdir=target src/handlers/*.ts",
		"lint": "eslint src/**/*.ts",
		"package": "pnpm type-check && pnpm lint && pnpm check-formatting && pnpm test && pnpm build && cd target; zip -qr zuora-salesforce-link-remover.zip ./*.js",
		"type-check": "tsc --noEmit",
		"check-formatting": "prettier --check **.ts",
		"fix-formatting": "prettier --write **.ts"
	},
	"dependencies": {
		"@aws-sdk/client-secrets-manager": "^3.654.0",
		"aws-lambda": "^1.0.7",
		"aws-sdk-client-mock": "^3.0.1",
		"zod": "^3.23.8"
	},
	"devDependencies": {
		"@types/aws-lambda": "^8.10.142"
	}
}
