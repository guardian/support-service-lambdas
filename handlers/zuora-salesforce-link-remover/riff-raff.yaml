stacks:
  - membership
regions:
  - eu-west-1
allowedStages:
  - CODE
  - PROD
deployments:
  zuora-salesforce-link-remover-cloudformation:
    type: cloud-formation
    app: zuora-salesforce-link-remover
    parameters:
      templateStagePaths:
        CODE: zuora-salesforce-link-remover-CODE.template.json
        PROD: zuora-salesforce-link-remover-PROD.template.json
  zuora-salesforce-link-remover:
    type: aws-lambda
    parameters:
      fileName: zuora-salesforce-link-remover.zip
      bucketSsmLookup: true
      prefixStack: false
      functionNames:
        - zuora-salesforce-link-remover-get-billing-accounts-
        - zuora-salesforce-link-remover-update-billing-accounts-in-zuora-
    dependencies: [zuora-salesforce-link-remover-cloudformation]
