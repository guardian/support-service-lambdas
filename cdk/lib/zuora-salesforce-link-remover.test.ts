import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { ZuoraSalesforceLinkRemover } from './zuora-salesforce-link-remover';

describe('The zuora-salesforce-link-remover stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const codeStack = new ZuoraSalesforceLinkRemover(app, 'zuora-salesforce-link-remover-CODE', {
			stack: 'membership',
			stage: 'CODE',
		});
		const prodStack = new ZuoraSalesforceLinkRemover(app, 'zuora-salesforce-link-remover-PROD', {
			stack: 'membership',
			stage: 'PROD',
		});

		expect(Template.fromStack(codeStack).toJSON()).toMatchSnapshot();
		expect(Template.fromStack(prodStack).toJSON()).toMatchSnapshot();
	});
});
