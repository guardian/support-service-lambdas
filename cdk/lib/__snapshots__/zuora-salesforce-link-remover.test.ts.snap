// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The zuora-salesforce-link-remover stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuLambdaFunction",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "getbillingaccountslambdaEDDA46BF": {
      "DependsOn": [
        "getbillingaccountslambdaServiceRoleDefaultPolicyEEDA5C9B",
        "getbillingaccountslambdaServiceRoleD8FEBB2C",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "membership/CODE/zuora-salesforce-link-remover/zuora-salesforce-link-remover.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "zuora-salesforce-link-remover",
            "STACK": "membership",
            "STAGE": "CODE",
          },
        },
        "FunctionName": "zuora-salesforce-link-remover-get-billing-accounts-CODE",
        "Handler": "getBillingAccounts.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "getbillingaccountslambdaServiceRoleD8FEBB2C",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "zuora-salesforce-link-remover",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "getbillingaccountslambdaServiceRoleD8FEBB2C": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "zuora-salesforce-link-remover",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "getbillingaccountslambdaServiceRoleDefaultPolicyEEDA5C9B": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "secretsmanager:GetSecretValue",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:secretsmanager:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":secret:DEV/Salesforce/ConnectedApp/AwsConnectorSandbox-oO8Phf",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:secretsmanager:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":secret:DEV/Salesforce/User/MembersDataAPI-yaO9GU",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/membership/CODE/zuora-salesforce-link-remover/zuora-salesforce-link-remover.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/membership/zuora-salesforce-link-remover",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/membership/zuora-salesforce-link-remover/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "getbillingaccountslambdaServiceRoleDefaultPolicyEEDA5C9B",
        "Roles": [
          {
            "Ref": "getbillingaccountslambdaServiceRoleD8FEBB2C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;

exports[`The zuora-salesforce-link-remover stack matches the snapshot 2`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuLambdaFunction",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "getbillingaccountslambdaEDDA46BF": {
      "DependsOn": [
        "getbillingaccountslambdaServiceRoleDefaultPolicyEEDA5C9B",
        "getbillingaccountslambdaServiceRoleD8FEBB2C",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "membership/PROD/zuora-salesforce-link-remover/zuora-salesforce-link-remover.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "zuora-salesforce-link-remover",
            "STACK": "membership",
            "STAGE": "PROD",
          },
        },
        "FunctionName": "zuora-salesforce-link-remover-get-billing-accounts-PROD",
        "Handler": "getBillingAccounts.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "getbillingaccountslambdaServiceRoleD8FEBB2C",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "zuora-salesforce-link-remover",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "getbillingaccountslambdaServiceRoleD8FEBB2C": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "zuora-salesforce-link-remover",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "getbillingaccountslambdaServiceRoleDefaultPolicyEEDA5C9B": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "secretsmanager:GetSecretValue",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:secretsmanager:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":secret:DEV/Salesforce/ConnectedApp/AwsConnectorSandbox-oO8Phf",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:secretsmanager:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":secret:DEV/Salesforce/User/MembersDataAPI-yaO9GU",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/membership/PROD/zuora-salesforce-link-remover/zuora-salesforce-link-remover.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/membership/zuora-salesforce-link-remover",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/membership/zuora-salesforce-link-remover/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "getbillingaccountslambdaServiceRoleDefaultPolicyEEDA5C9B",
        "Roles": [
          {
            "Ref": "getbillingaccountslambdaServiceRoleD8FEBB2C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
